covariance/covarianceBase.cpp

```diff
@@ -1393,125 +1389,266 @@ void covarianceBase::updateThrowMatrix(TMatrixDSym *cov){
+  outMeanVec->Write(systematicName+"_mean_vec");
+  outFile->Close();
+  delete outFile;
+}
 
-  TMatrixDSym* tmp_throwMatrix = (TMatrixDSym*)throwMatrixFile->Get(throwMatrixName);
-  TVectorD* tmp_meansvec = (TVectorD*)throwMatrixFile->Get(meansVectorName);
 
-  if(!tmp_throwMatrix){
-    std::cerr<<"ERROR : Couldn't find throw matrix "<<throwMatrixName<<" in "<<throwMatrixFileName<<std::endl;
-    std::cerr<<__FILE__<<" : "<<__LINE__<<std::endl;
+void covarianceBase::setAdaptionDefaults(){
+  // Puts adaptive MCMC default attributes somewhere obvious
+  use_adaptive          = false;
+  total_steps           = 0;
+  start_adaptive_throw  = 0;
+  start_adaptive_update = 0;
+  end_adaptive_update   = 1;
+  adaptive_update_step  = 1000;
+}
+
+// HW : Here be adaption
+void covarianceBase::initialiseAdaption(manager* fitMan){
+  /*
+    HW: Idea is that adaption can simply read the YAML config
+    Options :
+            External Info:
+              * UseExternalMatrix [bool]     :    Use an external matrix
+              * ExternalMatrixFileName [str] :    Name of file containing external info
+              * ExternalMatrixName [str]     :    Name of external Matrix
+              * ExternalMeansName [str]      :    Name of external means vector [for updates]
+
+            General Info:
+              * DoAdaption [bool]            :    Do we want to do adaption?
+              * AdaptionStartThrow [int]     :    Step we start throwing adaptive matrix from
+              * AdaptionEndUpdate [int]      :    Step we stop updating adaptive matrix
+              * AdaptionStartUpdate [int]    :    Do we skip the first N steps?
+              * AdaptionUpdateStep [int]     :    Number of steps between matrix updates
+              * Adaption blocks [vector<vector<int>>] : Splits the throw matrix into several block matrices
+ */
+
+  // need to cast matrixName to string
+  std::string matrixName_str(matrixName);
+  
+  //  setAdaptionDefaults();
+
+  if(!CheckNodeExists<std::string>(fitMan->raw(), "AdaptionOptions", "Covariance", matrixName_str)){
+    MACH3LOG_WARN("Adaptive Settings not found for {}, this is fine if you don't want adaptive MCMC", matrixName_str);
+    return;
+  }
+
+  // We"re going to grab this info from the YAML manager
+  if(!fitMan->raw()["AdaptionOptions"]["Covariance"][matrixName_str]["DoAdaption"].as<bool>()) {
+    MACH3LOG_INFO("Not using adaption for {}", matrixName_str);
+    return;
+  }
+
+  // Now we read the general settings [these SHOULD be common across all matrices!]
+  start_adaptive_throw  = fitMan->raw()["AdaptionOptions"]["Settings"]["AdaptionStartThrow"].as<int>();
+  start_adaptive_update =  fitMan->raw()["AdaptionOptions"]["Settings"]["AdaptionStartUpdate"].as<int>();
+  end_adaptive_update   =  fitMan->raw()["AdaptionOptions"]["Settings"]["AdaptionEndUpdate"].as<int>();
+  adaptive_update_step  =  fitMan->raw()["AdaptionOptions"]["Settings"]["AdaptionUpdateStep"].as<int>();
```

Comment:
Can you use `GetFromManager ` function to set default values and avoid crushes due to bad yaml?